src/scanner/mod.rs

```diff
@@ -100,41 +103,59 @@ impl Scanner {
     ///
     /// Note: `self` must contain `self.ip`.
     async fn scan_socket(&self, socket: SocketAddr) -> io::Result<SocketAddr> {
-        match self.connect(socket).await {
-            Ok(x) => {
-                debug!(
-                    "Connection was successful, shutting down stream {}",
-                    &socket
-                );
-                match x.shutdown(Shutdown::Both) {
-                    Err(e) => debug!("Shutdown stream error {}", &e),
-                    _ => {}
-                }
-                if !self.greppable {
-                    if self.accessible {
-                        println!("Open {}", socket.to_string());
-                    } else {
-                        println!("Open {}", socket.to_string().purple());
+        let mut ret = Err(io::Error::new(
```

Comment:
This variable is not needed if what I propose in the other comments is implemented. However, the default error still has to be returned if `self.tries` is 0, but you can move that behind the for loop, which will also avoid the allocation with `format!`, if it is not used.

Reply:
If `self.tries` is 0 it means the user gives 0 so I can modify the value when assigning it to `Scanner` and printing an information message.

Reply:
Even if this method never gets called with `self.tries == 0`, there has to be a return value for when it _would_ be 0. Since the error return type contains a `String` and it is possibe that you never enter the loop,  you will have to have some code that creates a string outside the loop. My point is to have that allocation be after the loop, because it is not needed otherwise.

Reply:
`cargo build` throws me an error if I did not set `ret` variable so I corrected this (even though in this case if we arrive after the loop the variable would have been set into the loop).